core/src/main/kotlin/org/coepi/api/v4/dao/TCNReportsDao.kt

```diff
@@ -22,40 +26,63 @@ class TCNReportsDao {
+     *
+     * @param timestamp the "created at" timestamp for this record
+     * @param reportData the payload for this report
+     */
+    fun addReport(
+        timestamp: Instant,
+        reportData: ByteBuffer
+    ): TCNReportRecord {
+        require(reportData.capacity() > 0) { "reportData cannot be empty" }
 
-        val reportId = generateReportId(date, intervalNumber)
+        val reportId = timestamp.toReportId()
         val randomId = UUID.randomUUID().toString()
-        val reportRecord = TCNReportRecord(reportId, randomId, timestamp, reportData.array())
-        this.dynamoMapper.save(reportRecord)
-        return reportRecord
-    }
 
-    fun queryReports(date: LocalDate, intervalNumber: Long): List<TCNReportRecord> {
-        val reportId = generateReportId(date, intervalNumber)
-        val queryExpression = DynamoDBQueryExpression<TCNReportRecord>()
-        queryExpression.keyConditionExpression = "reportId = :val1"
+        val record = TCNReportRecord(
+            reportId = reportId,
+            randomId = randomId,
+            reportTimestamp = timestamp.toEpochMilli(),
+            report = reportData.array()
+        )
+
+        dynamoMapper.save(record)
+
+        return record
+    }
 
-        val attributeValueMap = HashMap<String, AttributeValue>()
-        attributeValueMap[":val1"] = AttributeValue().withS(reportId)
-        queryExpression.expressionAttributeValues = attributeValueMap
+    fun queryReports(date: LocalDate, intervalNumber: Long): List<TCNReportRecord> =
+        getRecordsByReportId(
+            generateReportId(date, intervalNumber)
+        )
 
-        val outputList = mutableListOf<TCNReportRecord>()
-        var lastEvalKey: Map<String, AttributeValue>? = null
+    fun queryReports(intervalNumber: Long): List<TCNReportRecord> =
+        getRecordsByReportId(
+            reportIdFromInterval(intervalNumber)
+        )
 
-        do {
-            queryExpression.exclusiveStartKey = lastEvalKey
-            val pageOutput = dynamoMapper.queryPage(TCNReportRecord::class.java, queryExpression)
-            outputList.addAll(pageOutput.results)
-            lastEvalKey = pageOutput.lastEvaluatedKey
-        } while (lastEvalKey != null)
 
-        return outputList
+    private fun getRecordsByReportId(reportId: String): List<TCNReportRecord> {
+        val results = dynamoMapper.query(
+            DynamoDBQueryExpression<TCNReportRecord>()
+                .withKeyConditionExpression("reportId = :reportId")
+                .withExpressionAttributeValues(
+                    mapOf(":reportId" to AttributeValue(reportId))
+                )
+        )
+        results.loadAllResults()
+        return results
     }
-}
\ No newline at end of file
+}
+
+private fun reportIdFromInterval(intervalNumber: Long): String {
+    val timestamp = Instant.ofEpochMilli(intervalNumber * Intervals.INTERVAL_LENGTH_MS)
+
+    return "${timestamp.toUtcLocalDate()}:$intervalNumber"
```

Comment:
@bhushanRamnani What's the purpose of using the date as a prefix? Why not just use the interval number directly?

Seems like in an interval-focused world, we're just going to end up computing the date from the interval number most of the time anyway.

Reply:
The original purpose of date was to provide backwards compatibility for querying if the interval length changes in future. All reports stored using the older interval numbers could be queried using the older interval length for those dates and we wouldn't need to re-shard all existing reports in new interval buckets.

I realized that we can achieve the same backwards-compatibility by storing reports key-ed on ```intervalNumber:intervalLength``` combination. That's precisely the change I've made. So I think this PR is not needed. Would you like to implement the gaen endpoint instead ?